### PR TITLE
[REVIEW] Fix incorrect output from averages with filters in partial only mode

### LINKED ISSUE #155: [BUG] Incorrect output from averages with filters in partial only mode
**Describe the bug**
When using filters on averages with conf set to run only the partial mode on the gpu, the result can be different from the CPU.

**Steps/Code to reproduce bug**
Adding the following test to `hash_aggregate_test.py` shows the failure behaviour.
```
@ignore_order
@allow_non_gpu(
    'HashAggregateExec', 'AggregateExpression',
    'AttributeReference', 'Alias', 'Sum', 'Count', 'Max', 'Min', 'Average', 'Cast',
    'KnownFloatingPointNormalized', 'NormalizeNaNAndZero', 'GreaterThan', 'Literal', 'If',
    'EqualTo', 'First', 'SortAggregateExec', 'Coalesce')
@pytest.mark.parametrize('data_gen', get_params(_init_list_no_nans,
    marked_params=params_markers_for_avg_sum), ids=idfn)
def test_hash_multiple_filters_fail(data_gen):
    df = with_cpu_session(
        lambda spark : gen_df(spark, data_gen, length=100))
    df.createOrReplaceTempView("hash_agg_table")
    assert_gpu_and_cpu_are_equal_collect(
        lambda spark: spark.sql(
            'select avg(b) filter (where b > 20)' +
            'from hash_agg_table group by a'),
        conf=_no_nans_float_conf_partial)
```
**Expected behavior**
cpu result = `266969243.5`
gpu result = `None`

### PR DESCRIPTION
**Pull request by kuhushukla**
Fixes #155 .

This is WIP for a couple reasons and also as I want to get more comments on the approach shown in the PR.

As the associated issue explains in some detail, how we are ending up with nulls being sent down to the CPU's final aggregation. The fix tries to circumvent this by special casing how averages are handled by the case-when of the filter. One thing to note is, we need this special case since we don't have a clean way to pass the filter down to the `GpuAverage` DeclarativeAggregate itself. When a `null` comes into the filter we want it not pass down any further but instead default to `(0.0, 0)`. This seems to be specific for averages while other aggregates like `count` require nulls. I had to use another case-when with `isnotnull` to make this happen. I have some concerns on the else condition added in this PR and would like to know what others think is a _better_ way to do this. I tried some games with having initialValues be reused in the filter it was not straight forward what do to with their data types. At the end, I am also open to use this PR to morph the solution to put the hammer on filters+avg+partial_only_conf to fall back on the CPU, but that is not my first preference. Tagging @abellina for some discussion to get this PR going.

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala
Comment (kuhushukla) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:83]: Can do an instance of, agree. Will update.

Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:83]: or a `match` :) "scalafy!"

Comment (kuhushukla) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:83]: Yes! :) Have been in java land recently for too long. Will update here post some re-runs of tests.

**Comment by kuhushukla** [comment]
Addressed review comments @abellina and additionally took out a test that was redundant after this fix. Please take a look.

**Comment by kuhushukla** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:83]: why do we need the `asInstanceOf` here? Just curious. Maybe we should blow up if `origAggregateFunction` is not a `GpuDeclarativeAggregate`, but I think it'll still match without it being a declarative agg.

Comment (kuhushukla) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala]: my bad, debugging artifact. Removed now.

**Review by abellina — DISMISSED** [review]

**Comment by abellina** [comment]
build

**Comment by kuhushukla** [comment]
Waiting for approvals and CI, after which, I will rebase-squash and sign off.

**Comment by kuhushukla** [comment]
Squashed and signed off, re-runing CI

**Comment by kuhushukla** [comment]
build

**Pull request merged**
